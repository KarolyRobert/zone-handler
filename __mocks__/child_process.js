'use strict';

const { EventEmitter, Readable, Writable } = require('stream');
const child_process = jest.createMockFromModule('child_process');

class WriteMemory extends Writable {
    constructor() {
      super();
      this.buffer = "";
    }
  
    _write(chunk, _, next) {
      this.buffer += chunk;
      next();
    }
  
    reset() {
      this.buffer = "";
    }
  }

child_process._current;

child_process.spawn = jest.fn(() => {
    const cprocess = new EventEmitter();
    cprocess.stdin = new WriteMemory();
    cprocess.stdout = new Readable();
    cprocess.stderr = new Readable();



    cprocess.stdout._read = cprocess.stderr._read = (size) => {};
   
    child_process._current = cprocess;
    return cprocess;
});

module.exports = child_process;