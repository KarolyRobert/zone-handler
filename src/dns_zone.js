import nsupdate from './util/nsupdate';
import { transferZone } from './util/utils';
import dns_record from './dns_record';

export default function dns_zone(request){ // {zone,server,}
    let list = [];
    let hashtable = {};

    const transferRecord = record => {
        let new_record = dns_record(record,request.zone);
        if(hashtable[new_record.hash] === undefined){
            hashtable[new_record.hash] = list.length;
            list.push(new_record);
        }
    }

    const nsupdateCommands = command => {
        if(request.updateKey){
        return `zone ${request.zone}\n` +
            `key ${request.updateKey.algorithm}:${request.updateKey.name} ${request.updateKey.secret}\n` +
            `${command}send\n`;
        }else{
            throw new Error('Missing updateKey!');
        }
    };

    const getRecords = () => {
        return list.map(record => {
            return {
                hash:record.hash,
                record:record.record
            }
        });
    }

    const command = (type,hash,record) => {
        let index;
        switch(type){
            case 'add':
                let new_record = dns_record(record,request.zone);
                if(hashtable[new_record.hash] === undefined){
                    return nsupdateCommands(new_record.add_command());
                }else{
                    throw new Error(`Such record already is in the ${request.zone} zone!`);
                }
            case 'update':
                index = hashtable[hash];
                if(index > -1){
                    let old_record = list[index];
                    let new_record = dns_record(record,request.zone);
                    if(!hashtable[new_record.hash]){
                        return nsupdateCommands(old_record.update_command(new_record));
                    }else{
                        throw new Error(`Such record already is in the ${request.zone} zone!`);
                    }
                }else{
                    throw new Error(`Not such record in ${request.zone} zone!`);
                }
            case 'delete':
                index = hashtable[hash];
                if(index > -1){
                    let deleted_record = list[index];
                    return nsupdateCommands(deleted_record.delete_command());
                }else{
                    throw new Error(`Not such record in ${request.zone} zone!`);
                } 
        }
    };



    const add_record = record => {
        try{
            let nsupdateCommand = command('add',null,record);
            return nsupdate(nsupdateCommand);
        }catch(err){
            Promise.reject(err);
        }
    }
    const update_record = (hash,record) => {
        try{
            let nsupdateCommand = command('update',hash,record);
            return nsupdate(nsupdateCommand);
        }catch(err){
            Promise.reject(err);
        }
    }
    const delete_record = hash => {
        try{
            let nsupdateCommand = command('delete',hash);
            return nsupdate(nsupdateCommand);
        }catch(err){
            Promise.reject(err);
        }
    }

    return new Promise((resolve,reject) => {
        transferZone(request,transferRecord).then(() => {
            resolve({
                getRecords,
                add : add_record,
                update : update_record,
                delete : delete_record
            });
        },err => reject(err));
    });
}
