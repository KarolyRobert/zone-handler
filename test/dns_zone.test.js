import child_process from 'child_process';
import dns_zone from '../src/dns_zone';

jest.mock('child_process');

const digResponse = '; <<>> DiG 9.11.5-P4-5.1+deb10u5-Raspbian <<>> @ns0.hobbyfork.com hobbyfork.com axfr -k tsig.key\n'+
'; (1 server found)\n'+
';; global options: +cmd\n'+
'hobbyfork.com.		21600	IN	SOA	ns0.hobbyfork.com. hostmaster.hobbyfork.com. 2021081502 43200 7200 2419200 600\n'+
'hobbyfork.com.		604800	IN	NS	ns0.hobbyfork.com.\n'+
'hobbyfork.com.		604800	IN	NS	ns1.hobbyfork.com.\n'+
'hobbyfork.com.		604800	IN	A	92.118.27.14\n'+
'hobbyfork.com.		604800	IN	RRSIG	A 13 2 604800 20210823025040 20210809170317 22370 hobbyfork.com. rrMcsVrp2rIzmrz7EgTNa2vsJyZjdDLPOaMjbufQIt6V+A 7GSEtB0SuW92GZGg==\n'+
'hobbyfork.com.		604800	IN	MX	10 mail.hobbyfork.com.\n'+
'hobbyfork.com.		86400	IN	TXT	"v=spf1 mx ~all"\n'+
'hobbyfork.com.		604800	IN	DNSKEY	257 3 13 xDMv4a6Cfgk0al1GfePls+fslmVfsFQdRegZSAvyj4js6Y8URqpjPfMCk 1Qy2UNW/5PLwN3pJORaXxpUpN4uCBQ==\n'+
'hobbyfork.com.		0	IN	NSEC3PARAM 1 0 10 30359FDCFA1D73EF\n'+
'hobbyfork.com.		3600	IN	CDS	22370 13 2 C74396E179A86BC6E3ACA45E283A168DA5E7707083632F88B6AE2548 3C47E30C\n'+
'hobbyfork.com.		3600	IN	CDNSKEY	257 3 13 xDMv4a6Cfgk0al1GfePl+fslmVfsFQdRegZ6Y8URqpjPfMCk 1Qy2UNW/5PLwN3pxpUpN4uCBQ==\n'+
'hobbyfork.com.		0	IN	TYPE65534 \# 5 0D57620001\n'+
'_dmarc.hobbyfork.com.	86400	IN	TXT	"v=DMARC1; p=reject; pct=100; fo=1; rua=mailto:postmaster@hobbyfork.com"\n'+
'default._domainkey.hobbyfork.com. 86400	IN TXT	"v=DKIM1; h=sha256; k=rsa; " "p=MIIBIjggdANBgdkLqmTMaE89iRTpZd8pyHvsBWC9UTKPMyGYg//ZeW7KEHhQuQkSvKO3HsFN2pzeQI+UGszHviykKXz3SxReHSscolEEmxnckc8mtK78DO3KFbg6lp" "aqUNjhhOnVEvsYpwC6FIw09ZpfWghemQ9JDPI+xVpapuVvzvUbwIDAQAB"\n'+
'mail.hobbyfork.com.	604800	IN	A	92.118.27.14\n'+
'ns0.hobbyfork.com.	604800	IN	A	45.148.28.74\n'+
'ns1.hobbyfork.com.	604800	IN	A	92.118.27.14\n'+
'www.hobbyfork.com.	30	IN	A	37.234.91.82\n'+
'2NHI2QRRGJ21UFACQGRECNRF7P6EF7S2.hobbyfork.com.	600 IN NSEC3 1 1 10 30359FDCFA1D73EF 3OG57RMC11QJC4SH0SEN6RHEB30EPMOH CNAME RRSIG\n'+
'hobbyfork.com.		21600	IN	SOA	ns0.hobbyfork.com. hostmaster.hobbyfork.com. 2021081502 43200 7200 2419200 600\n'+
'tsig.			0	ANY	TSIG	hmac-sha256. 1629138967 300 32 0snJBxPUhMsKUZ9Bp+HCyTldosvSvhNO+HEsgASSQS0M= 55494 NOERROR 0 \n'+
';; Query time: 80 msec\n'+
';; SERVER: 45.148.28.74#53(45.148.28.74)\n'+
';; WHEN: h aug 16 20:36:07 CEST 2021\n'+
';; XFR size: 66 records (messages 1, bytes 5844)\n';


describe('dns_zone', () => {

    test('dns_zone promise an object with getRecords, and command properties.',done => {
        expect.assertions(5);
        dns_zone({
            zone:'hobbyfork.com',
            server:'ns0.hobbyfork.com',
            updateKey:{
                name:'update.key',
                algorithm:'hmac-sha256',
                secret:'asdfksf8s6s875g765'
            },
            transferKey:{
                name:'tsig',
                algorithm:'hmac-sha256',
                secret:'a765hs6h7sdh75g765'
            }
        }).then(zone => {
            let hases = zone.getRecords().map(record => record.hash);
            expect(zone.getRecords().length).toBe(12);
            expect(hases).toEqual([
                "9aa25f49e7c4e9d05f30862818f57717",
                "6f3a8a9d844888acc7cd3bb3e88c5767",
                "8b946252701c4b2d3fc3d1b9cc1a7d8b",
                "e0832e4c3c82cd6b9a4a01f7a509e62a",
                "75e21d6ec44bf09aba586b0faafa7601",
                "fcad8a0f3a6a84f80b1653de77ea2e0b",
                "7bff7fabba00a16fa3cb500592c11095",
                "b81fe34e0b805d0b13f449173b03b746",
                "81a4c948c1b2e6dbfe0d0e72012a15ca",
                "1357d488cbc526fe2f351b94b4c8b382",
                "a453d4c826191769b371e3f18efb172f",
                "2a35bcda975bb00eaf1a26278314d905",
            ]);
            expect(zone.command('add',null,'home.hobbyfork.com. 21600 IN A 192.168.1.6')).toBe(
                'zone hobbyfork.com\n'+
                'key hmac-sha256:update.key asdfksf8s6s875g765\n'+
                'update add home.hobbyfork.com. 21600 A 192.168.1.6\n'+
                'send\n');
            expect(zone.command('update','2a35bcda975bb00eaf1a26278314d905','home.hobbyfork.com. 21600 IN A 192.168.1.6')).toBe(
                'zone hobbyfork.com\n'+
                'key hmac-sha256:update.key asdfksf8s6s875g765\n'+
                'update delete www.hobbyfork.com. 30 A 37.234.91.82\n'+
                'update add home.hobbyfork.com. 21600 A 192.168.1.6\n'+
                'send\n');
            expect(zone.command('delete','2a35bcda975bb00eaf1a26278314d905')).toBe(
                'zone hobbyfork.com\n'+
                'key hmac-sha256:update.key asdfksf8s6s875g765\n'+
                'update delete www.hobbyfork.com. 30 A 37.234.91.82\n'+
                'send\n');
            done();
        },err => {
            expect(err).toBe('error');
            done();
        });
        child_process._current.stdout.push(digResponse);
        process.nextTick(() => child_process._current.emit('close',0));
    });

});



