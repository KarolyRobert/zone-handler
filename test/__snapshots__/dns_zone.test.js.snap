// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dns_zone dns_zone default use case. 1`] = `
Array [
  Object {
    "hash": "9aa25f49e7c4e9d05f30862818f57717",
    "record": Object {
      "data": "ns0.hobbyfork.com. hostmaster.hobbyfork.com. 2021081502 43200 7200 2419200 600",
      "name": "hobbyfork.com.",
      "ttl": 21600,
      "type": "SOA",
    },
  },
  Object {
    "hash": "6f3a8a9d844888acc7cd3bb3e88c5767",
    "record": Object {
      "data": "ns0.hobbyfork.com.",
      "name": "hobbyfork.com.",
      "ttl": 604800,
      "type": "NS",
    },
  },
  Object {
    "hash": "8b946252701c4b2d3fc3d1b9cc1a7d8b",
    "record": Object {
      "data": "ns1.hobbyfork.com.",
      "name": "hobbyfork.com.",
      "ttl": 604800,
      "type": "NS",
    },
  },
  Object {
    "hash": "9761043884602298e407bc24fc0b4eae",
    "record": Object {
      "data": "92.111.27.14",
      "name": "hobbyfork.com.",
      "ttl": 604800,
      "type": "A",
    },
  },
  Object {
    "hash": "75e21d6ec44bf09aba586b0faafa7601",
    "record": Object {
      "data": "10 mail.hobbyfork.com.",
      "name": "hobbyfork.com.",
      "ttl": 604800,
      "type": "MX",
    },
  },
  Object {
    "hash": "fcad8a0f3a6a84f80b1653de77ea2e0b",
    "record": Object {
      "data": "v=spf1 mx ~all",
      "name": "hobbyfork.com.",
      "ttl": 86400,
      "type": "TXT",
    },
  },
  Object {
    "hash": "7bff7fabba00a16fa3cb500592c11095",
    "record": Object {
      "data": "v=DMARC1; p=reject; pct=100; fo=1; rua=mailto:postmaster@hobbyfork.com",
      "name": "_dmarc.hobbyfork.com.",
      "ttl": 86400,
      "type": "TXT",
    },
  },
  Object {
    "hash": "b81fe34e0b805d0b13f449173b03b746",
    "record": Object {
      "data": "v=DKIM1; h=sha256; k=rsa; p=MIIBIjggdANBgdkLqmTMaE89iRTpZd8pyHvsBWC9UTKPMyGYg//ZeW7KEHhQuQkSvKO3HsFN2pzeQI+UGszHviykKXz3SxReHSscolEEmxnckc8mtK78DO3KFbg6lpaqUNjhhOnVEvsYpwC6FIw09ZpfWghemQ9JDPI+xVpapuVvzvUbwIDAQAB",
      "name": "default._domainkey.hobbyfork.com.",
      "ttl": 86400,
      "type": "TXT",
    },
  },
  Object {
    "hash": "81a4c948c1b2e6dbfe0d0e72012a15ca",
    "record": Object {
      "data": "92.118.27.14",
      "name": "mail.hobbyfork.com.",
      "ttl": 604800,
      "type": "A",
    },
  },
  Object {
    "hash": "1357d488cbc526fe2f351b94b4c8b382",
    "record": Object {
      "data": "45.148.28.74",
      "name": "ns0.hobbyfork.com.",
      "ttl": 604800,
      "type": "A",
    },
  },
  Object {
    "hash": "a453d4c826191769b371e3f18efb172f",
    "record": Object {
      "data": "92.118.27.14",
      "name": "ns1.hobbyfork.com.",
      "ttl": 604800,
      "type": "A",
    },
  },
  Object {
    "hash": "2a35bcda975bb00eaf1a26278314d905",
    "record": Object {
      "data": "37.234.91.82",
      "name": "www.hobbyfork.com.",
      "ttl": 30,
      "type": "A",
    },
  },
]
`;

exports[`dns_zone dns_zone default use case. 2`] = `
"zone hobbyfork.com
key hmac-sha256:update.key asdfksf8s6s875g765
update add home.hobbyfork.com. 21600 A 192.168.1.6
send
"
`;

exports[`dns_zone dns_zone default use case. 3`] = `
"zone hobbyfork.com
key hmac-sha256:update.key asdfksf8s6s875g765
update delete www.hobbyfork.com. 30 A 37.234.91.82
update add www.hobbyfork.com. 21600 A 34.123.39.11
send
"
`;

exports[`dns_zone dns_zone default use case. 4`] = `
"zone hobbyfork.com
key hmac-sha256:update.key asdfksf8s6s875g765
update delete www.hobbyfork.com. 30 A 37.234.91.82
send
"
`;
